[custom_op] triton_op API V0

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130637

This is the initial version of an API to create custom operators whose
implementations are backed by triton kernels. While user-defined triton
kernels work out-of-the-box with triton kernels, you may wish to
construct a custom operator if you need to compose with other PyTorch
subsystems, like Tensor subclasses or vmap.

I'm hoping to get design feedback on this and ship it so that we can
begin experimenting with customers.

Test Plan:
- new tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

zou3519 committed:
[custom_op] triton_op API V0

This is the initial version of an API to create custom operators whose
implementations are backed by triton kernels. While user-defined triton
kernels work out-of-the-box with triton kernels, you may wish to
construct a custom operator if you need to compose with other PyTorch
subsystems, like Tensor subclasses or vmap.

I'm hoping to get design feedback on this and ship it so that we can
begin experimenting with customers.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] triton_op API V0"


This is the initial version of an API to create custom operators whose
implementations are backed by triton kernels. While user-defined triton
kernels work out-of-the-box with triton kernels, you may wish to
construct a custom operator if you need to compose with other PyTorch
subsystems, like Tensor subclasses or vmap.

I'm hoping to get design feedback on this and ship it so that we can
begin experimenting with customers.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented on torch/_library/triton.py:
```diff
@@ -0,0 +1,192 @@
+        return dec(fn)
+
+
+def capture_triton(triton_kernel: Callable, /) -> Callable:
```
This was moved from torch/_higher_order_ops/triton_kernel_wrap.py

zou3519 commented on torch/_library/triton.py:
```diff
@@ -0,0 +1,192 @@
+            in PyTorch subsystems (e.g. torch.export, FX graphs).
+            To avoid name collisions, please use your project name as the namespace;
+            e.g. all custom ops in pytorch/fbgemm use "fbgemm" as the namespace.
+        mutates_args (Iterable[str] or "unknown"): The names of args that the function mutates.
+            This MUST be accurate, otherwise, the behavior is undefined. If "unknown",
+            it pessimistically assumes that all inputs to the operator are being mutated.
```
This is required right now, but in theory we can infer it or assume all functional and error out if we detect that isn't the case. I can handle that in a follow-up.

zou3519 commented on torch/_library/triton.py:
```diff
@@ -0,0 +1,192 @@
+
+    """
+
+    def dec(fn: Callable) -> Callable:
```
The interesting thing about `@triton_op` is that there is nothing triton-specific about it at the moment. This seems like more of a "how do we create a custom op that decomposes right before Inductor" type of deal. But I think triton kernels are big enough that they warrant their own high-level API.

albanD commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1071,17 +998,13 @@ def __getitem__(self, *args):
-        import torch._dynamo
-
-        if not is_fx_tracing() or torch._dynamo.is_compiling():
+        if False:
```
I guess these will be fixed?

albanD commented on torch/_library/triton.py:
```diff
@@ -0,0 +1,192 @@
+            with mode:
+                return fn(*args, **kwargs)
+
+        result.register_torch_dispatch(FunctionalTensorMode, functional_decomp)
```
Very cool!

albanD reviewed (approved):
Looks great!

zou3519 committed:
Update on "[custom_op] triton_op API V0"


This is the initial version of an API to create custom operators whose
implementations are backed by triton kernels. While user-defined triton
kernels work out-of-the-box with triton kernels, you may wish to
construct a custom operator if you need to compose with other PyTorch
subsystems, like Tensor subclasses or vmap.

I'm hoping to get design feedback on this and ship it so that we can
begin experimenting with customers.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge -f "unrelated failure"
